### ISSUE TITLE
Trying to overfit a minibatch with `SoftmaxLikelihood`

### ISSUE THREAD

**Issue by lopezpaz** [cornellius-gp/gpytorch#1667]
Hello,

I am trying to overfit a mini-batch of synthetic data for multi-class classification, with no luck. In fact, the los value barely moves, no matter how long I optimize. I attach the minimal failing example below. The version of `torch` is `1.8` and the one of `gpytorch` is `1.4.2`. What's wrong with my code? Thank you!

```python	
import torch
import gpytorch

class GP(gpytorch.models.ApproximateGP):
    def __init__(self, x, num_tasks):
        variational_distribution = \
            gpytorch.variational.CholeskyVariationalDistribution(len(x))

        variational_strategy = \
            gpytorch.variational.IndependentMultitaskVariationalStrategy(
                gpytorch.variational.VariationalStrategy(
                    self, x, variational_distribution),
                num_tasks=num_tasks)

        super(GP, self).__init__(variational_strategy)

        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel())

    def forward(self, x):
        return gpytorch.distributions.MultivariateNormal(
            self.mean_module(x), self.covar_module(x))


if __name__ == "__main__":
    num_tasks = 50
    x = torch.randn(32, 100)
    y = torch.randn(32).random_(num_tasks).long()

    model = GP(x, num_tasks)

    likelihood = gpytorch.likelihoods.SoftmaxLikelihood(
        num_classes=num_tasks, mixing_weights=False)

    loss = gpytorch.mlls.VariationalELBO(
        likelihood,
        model,
        num_data=len(x))

    optimizer = torch.optim.Adam(model.parameters(), lr=1e-2)

    for epoch in range(500):
        loss_value = -loss(model(x), y)

        optimizer.zero_grad()
        loss_value.backward()
        optimizer.step()

        accuracy = model(x).mean.argmax(1).eq(y).float().mean().item()
        print(loss_value.item(), accuracy)
```

**Comment by gpleiss** [comment]
Modify your `variational_distribution` to be:

```python
variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
    len(x), batch_shape=torch.Size([num_tasks])
)
```

Without the `batch_shape` argument, you're learning a shared set of variational parameters for all tasks.

**Comment by lopezpaz** [comment]
Thanks for the correction @gpleiss. However, I am still unable to learn. In an effort to make things better, I have implemented the median heuristic to choose the RBF kernel initial lengthscale, and the clustering heuristic to choose the initial inducing locations. I have set a fixed random seed to illustrate that the following script gets 0.0625 classification accuracy. Any ideas?

For comparison, a normal linear layer trained by empirical risk minimization nails this toy problem in 5 iterations.

Thank you for your help!

```python
import torch
import gpytorch
from sklearn import cluster

class GP(gpytorch.models.ApproximateGP):
    def __init__(self,
                 num_outputs,
                 some_features,
                 num_inducing_points=32):
        batch_shape = torch.Size([num_outputs])

        # clustering heuristic to determine the initial inducing locations
        initial_inducing_points = self.cluster_(
            some_features, num_inducing_points)

        variational_distribution = \
            gpytorch.variational.CholeskyVariationalDistribution(
                num_inducing_points, batch_shape=batch_shape)

        variational_strategy = \
            gpytorch.variational.IndependentMultitaskVariationalStrategy(
                gpytorch.variational.VariationalStrategy(
                    self, initial_inducing_points, variational_distribution),
                num_tasks=num_outputs)

        super(GP, self).__init__(variational_strategy)

        kernel = gpytorch.kernels.RBFKernel(batch_shape=batch_shape)

        # median heuristic to determine initial lengthscale
        ils = torch.pdist(some_features.detach().cpu()).mean().item()
        kernel.lengthscale = ils * torch.ones_like(kernel.lengthscale)

        self.mean_module = gpytorch.means.ConstantMean(batch_shape=batch_shape)
        self.covar_module = gpytorch.kernels.ScaleKernel(
            kernel, batch_shape=batch_shape)

    def cluster_(self, some_features, k):
        kmeans = cluster.MiniBatchKMeans(n_clusters=k, batch_size=k)
        kmeans.fit(some_features.detach().cpu())
        return torch.from_numpy(kmeans.cluster_centers_).float()

    def forward(self, x):
        return gpytorch.distributions.MultivariateNormal(
            self.mean_module(x), self.covar_module(x))
    
def main():
    torch.manual_seed(0)

    num_tasks = 250
    x = torch.randn(32, 100)
    y = torch.randn(32).random_(num_tasks)

    model = GP(num_tasks, x)

    likelihood = gpytorch.likelihoods.SoftmaxLikelihood(
        num_classes=num_tasks, mixing_weights=False)

    loss = gpytorch.mlls.VariationalELBO(
        likelihood,
        model,
        num_data=len(x))

    optimizer = torch.optim.SGD(model.parameters(), momentum=0.9, lr=1e-1)

    for epoch in range(500):
        loss_value = -loss(model(x), y)

        optimizer.zero_grad()
        loss_value.backward()
        optimizer.step()

        model.eval()
        accuracy = model(x).mean.argmax(1).eq(y).float().mean().item()
        print(loss_value.item(), accuracy)
        model.train()


if __name__ == "__main__":
    main()
```
